mapo/agents/mapo/losses.py

```diff
@@ -0,0 +1,125 @@
+"""Collection of loss functions in tensorflow."""
+import tensorflow as tf
+from tensorflow import keras
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.evaluation.postprocessing import Postprocessing
+
+
+def dynamics_mle_loss(batch_tensors, model):
+    """Compute dynamics loss via Maximum Likelihood Estimation."""
+    obs, actions, next_obs = (
+        batch_tensors[SampleBatch.CUR_OBS],
+        batch_tensors[SampleBatch.ACTIONS],
+        batch_tensors[SampleBatch.NEXT_OBS],
+    )
+    return -tf.reduce_mean(model.compute_states_log_prob(obs, actions, next_obs))
+
+
+def dynamics_pga_loss(batch_tensors, model, actor_loss, config):
+    """Compute Policy Gradient Aware dynamics loss"""
+    gmapo = tf.gradients(actor_loss, model.actor_variables)
+    flat_gmapo = tf.concat([tf.reshape(grad, [-1]) for grad in gmapo], axis=0)
```

Comment:
Maybe it would be better if we leave this concatenation step as an option to the kernel function. 
Since not all kernels are additive, I'm not sure this is the best way to go. You should be able to test both ways (e.g., the sum of the results of the kernels for each variable Vs. the kernel applied to the concatenated vector all at once) ... 

So, I think it would make a more flexible solution to pass to the kernel both lists of gradient vectors (aligned wrt to the actor variables) and let the kernel compute the scalar loss... ok?

Reply:
Makes sense, I'll implement it right away.